apps/designer/app/shared/form-utils/validation.ts

```diff
@@ -0,0 +1,79 @@
+import type { Fetcher } from "@remix-run/react";
+import { toast } from "@webstudio-is/design-system";
+import { useCallback, useState } from "react";
+import type { ZodError } from "zod";
+import { useOnFetchEnd } from "~/shared/fetcher";
+
+export type FetcherData<Payload> =
+  | ({ status: "ok" } & Payload)
+  | { status: "error"; errors: string | ZodError["formErrors"] };
+
+export const normalizeErrors = (
+  errors: string | ZodError["formErrors"]
+): ZodError["formErrors"] =>
+  typeof errors === "string"
+    ? { formErrors: [errors], fieldErrors: {} }
+    : errors;
+
+export const useFetcherErrors = <FieldName extends string>({
+  fetcher,
+  fieldNames,
+}: {
+  fetcher: Fetcher<FetcherData<unknown>>;
+  fieldNames?: readonly FieldName[];
+}): {
+  fieldErrors: ZodError["formErrors"]["fieldErrors"];
+  resetFieldError: (fieldName: FieldName) => void;
+} => {
+  const [fieldErrors, setFieldErrors] = useState<
+    ZodError["formErrors"]["fieldErrors"]
+  >({});
+
+  useOnFetchEnd(fetcher, (data) => {
+    if (data.status === "error") {
+      const errors = normalizeErrors(data.errors);
+      toastUnknownFieldErrors(errors, fieldNames);
+      setFieldErrors(errors.fieldErrors);
+    }
+  });
+
+  const resetFieldError = useCallback((fieldName: FieldName) => {
+    setFieldErrors(({ [fieldName]: _, ...rest }) => rest);
+  }, []);
+
+  return { fieldErrors, resetFieldError };
+};
+
+// Show a toast for each of formErrors
+// as well as fieldErrors which we cannot display near a corresponding field
+export const toastUnknownFieldErrors = (
+  errors: ZodError["formErrors"],
+  knownFields?: readonly string[]
+) => {
+  for (const message of errors.formErrors) {
+    toast.error(message);
+  }
+
+  if (knownFields !== undefined) {
+    for (const fieldName in errors.fieldErrors) {
+      if (knownFields.includes(fieldName) === false) {
+        for (const message of errors.fieldErrors[fieldName] as string[]) {
```

Comment:
also should be possible to not assert, need help?

Reply:
seems like you should be checking if fieldName is in fieldErrors instead of asserting it

Reply:
These `as` statements are fine IMO. I don't see how they can lead to bugs. But sure, if you or @Andarist knows how to improve this, let improve.

Reply:
The way I see it is you are asserting that those field errors for a certain field exist, but the code doenstn't prove that, its just telling compiler this is true.

You are probably checking it by checking knownFields and known fields automatically have keys in fieldErrors, but this logic is implicit and has no explicit verification. Sometimes its too hard to make it more obvious, so we do this type of stuff, but we should def. try to avoid it

Reply:
Yeah, this one was actually a potential bug, I misunderstood originally what TS complains about. Fixed.